### ISSUE TITLE
Verify command issue when executing command with a for /F statement

### ISSUE THREAD

**Issue by houghtonap** [MicrosoftDocs/windowsserverdocs#5495]
There is a subtle bug in the windows command processor with regard to the `verify` command. You will receive two different results when executing the command in a batch script vs. executing the command with a `for /F` statement in a batch script. The following batch script demonstrates the issue:

```cmd
@setlocal
  @prompt $N:$G$S
  verify
  for /F "usebackq tokens=3 delims=. " %%a in (` verify `) do echo verify=%%a=
  @prompt
@endlocal
@goto :EOF
```

which produces:

```txt
U:> verify
VERIFY is on.

U:> for /F "usebackq tokens=3 delims=. " %a in (` verify `) do echo verify=%a=

U:> echo verify=off=
verify=off=
```

Note that `verify` is on when the above batch script is run, but when trying to determine the state of `verify` with the `for /F` statement it says it's `off`. I suspect that this may be a result of the `for /F` statement executing the result in a different command processor where the default for `verify` is `off`. However, this seems to be an oddity with executing the `verify` command. When you do this for the `echo` command with the following similar batch script there are no issues:

```cmd
@setlocal
  @prompt $N:$G$S
  echo
  for /F "usebackq tokens=3 delims=. " %%a in (` echo `) do echo echo=%%a=
  @prompt
@endlocal
@goto :EOF
```

which correctly produces:

```txt
U:> echo
ECHO is on.

U:> for /F "usebackq tokens=3 delims=. " %a in (` echo `) do echo echo=%a=

U:> echo echo=on=
echo=on=
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 09f8b51a-8984-72d4-c041-9eb8320fef7b
* Version Independent ID: d18467dc-98bd-4fd4-ceca-6693d755d14a
* Content: [verify](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/verify)
* Content Source: [WindowsServerDocs/administration/windows-commands/verify.md](https://github.com/MicrosoftDocs/windowsserverdocs/blob/master/WindowsServerDocs/administration/windows-commands/verify.md)
* Product: **windows-server**
* Technology: **windows-commands**
* GitHub Login: @JasonGerend
* Microsoft Alias: **jgerend**

**Comment by houghtonap** [comment]
Note, this is also a problem with the `echo` command. I didn't formulate the test conditions appropriately in my original post of the `verify` issue. The following script modification shows the issue with the `echo` command too:

```cmd
@setlocal
  @prompt $N:$G$S
  echo off
  for /F "usebackq tokens=3 delims=. " %%a in (` echo `) do echo echo=%%a=
  @prompt
@endlocal
@goto :EOF
```

My initial post comments about the `echo` command were incorrect and test scenario was incorrect because the default for the `echo` command is `on`, thus when it is executed in the `for /F` statement it "appears" to match. However, the altered test scenario for the `echo` command first changes `echo` to be `off` before executing the `for /F` statement and it also shows two different results:

```cmd
U:> echo off
echo=on=
```

**Comment by illfated** [comment]
While not having researched this issue yet, I would like to mention that Windows batch scripts often start with `@Echo off` on the first line, although some scripts skip that step.

Would it be useful to discuss the implications of using or not using that line as the first line of the script?

**Comment by houghtonap** [comment]
@illfated Many scripts that are written incorrectly do the following:

```cmd
@echo off
  rem some commands that the batch script doesn't want echo'ed...
@echo on & goto :EOF
```

The problem with this strategy is that you have a 50/50 chance of returning `echo` to the correct state that your caller was using. Let's take two batch scripts:

`test-1.cmd`
```cmd
@echo off
setlocal
  set /A a= 0
  call test-2.cmd
  set /A a= a + 1
endlocal
goto :EOF
```

`test-2.cmd`
```cmd
@echo off
setlocal
  dir
endlocal
@echo on & goto :EOF
```

This results in the following `test-1.cmd` commands being echo'ed:

```cmd
  set /A a= a + 1
endlocal
goto :EOF
```
this is not what the `test-1.cmd` batch script was expecting.

However, you may protest, I want both scripts to not `echo` commands, there shouldn't be an `echo on` at the end of `test-2.cmd`. Let's make that change:

`test-2.cmd`
```cmd
@echo off
setlocal
  dir
endlocal
goto :EOF
```

and in fact when running `test-1.cmd` from a **command prompt** we only see the directory listing that was executed in `test-2.cmd`, but let's introduce another script and call it from the command prompt:

`test-0.cmd`
```cmd
echo on
setlocal
rem I really want these commands to be echo'ed
call test-1.cmd
for /D %%D in ( . ) do set lastDir=%%D
endlocal
```

hmm... that `for /D` command is not echo'ed and that's not what was expected by the `test-0.cmd` batch script.

Without knowing the state of your caller's `echo` or `verify` you are likely to introduce unintended side effects for other scripts. Blindly turning `echo` off in each script only works when **all** scripts expect `echo off`.

When you are developing a bunch of generic scripts where you don't know how they will be orchestrated together you cannot assume `echo off` is the correct choice.

**Comment by illfated** [comment]
Fair enough, although I thought it would be necessary to move the `echo on` line in `test-0.cmd` down and inside the `setlocal` subsection. I am not an expert on setlocal / endlocal usage, so I will wait for more feedback from any of the MS Docs team members, especially regarding the [`verify`](https://docs.microsoft.com/windows-server/administration/windows-commands/verify) page. While we're waiting, maybe you want to see what other knowledgeable users may have to say in the forum pages?

Disclaimer: I am not affiliated with Microsoft or the Microsoft Docs teams here on GitHub.

- TechNet forums: https://social.technet.microsoft.com/Forums/en-US/home
- MSDN: https://social.msdn.microsoft.com/Forums/en-US/home
- Microsoft Community: https://answers.microsoft.com/

**Comment by Xelu86** [comment]
#assign:xelu86

**Comment by Xelu86** [comment]
Hey @houghtonap,

Thanks again for contacting Microsoft. We value your feedback. 

This is a lot of testing you've gone through with this, commendable! As this is a short reference on the use of the `verify` command, we aren't able to discuss all the outputs produced in batch script making. GitHub isn't quite the right location if searching for technical support, feature requests, or the information you kindly provided as it is completely outside our scope to provide any insight. If there are any issues within the content itself, please let us know.

This information is beneficial and maybe a community expert can better assist as to why you are receiving this output with your scripts. Please open a ticket with [Microsoft Support](https://support.microsoft.com/ContactUs) to see if this has been brought up before with any potential solutions.

Illfated already provided our other tech support forums in the post above. As we're unable to action this due to its scope, we will now be closing this thread. 

Thank you and we do wish you a most pleasant day! 
#please-close
